From 4ecda064824ef5fb632e49ca46beb1f03300db19 Mon Sep 17 00:00:00 2001
From: Balamurugan Ramar <quic_bramar@quicinc.com>
Date: Thu, 10 Apr 2025 23:27:22 +0530
Subject: [PATCH] mac80211: QSDK wifi add MHz vendor IE support for STA mode

- Store Q-Q 240 MHz vendor IE in the bss data structure for STA mode
and add it in the Association Request frame sent to AP.
- Use same data to upgrade from 160 MHz to 320 MHz (with punturing
bitmap) while generating channel data for the link.
- Peer association parameters sent to firmware via WMI will reflect
these bandwidth changes as well.

Patch-dependency: 927-003-QSDK-wifi-mac80211-parse-and-store-240-MHz-vendor-IE.patch
Patch-work: None

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
Signed-off-by: Balamurugan Ramar <quic_bramar@quicinc.com>
---
 include/linux/ieee80211.h  |  7 +++++
 net/mac80211/ieee80211_i.h |  8 +++++-
 net/mac80211/mesh.c        |  2 +-
 net/mac80211/mlme.c        | 47 ++++++++++++++++++++++++++++-----
 net/mac80211/parse.c       |  2 +-
 net/mac80211/scan.c        |  7 +++++
 net/mac80211/tdls.c        |  2 +-
 net/mac80211/util.c        | 54 +++++++++++++++++++++++++++++++++++---
 8 files changed, 116 insertions(+), 13 deletions(-)

diff --git a/include/linux/ieee80211.h b/include/linux/ieee80211.h
index 262413d..e1fe430 100644
--- a/include/linux/ieee80211.h
+++ b/include/linux/ieee80211.h
@@ -2412,6 +2412,11 @@ struct ieee80211_eht_operation_info {
 	u8 optional[];
 } __packed;
 
+struct ieee80211_qcn_version_vendor_elem {
+	u8 version;
+	u8 subversion;
+}__packed;
+
 struct ieee80211_240mhz_vendor_oper {
 	u8 ccfs1;
 	u8 ccfs0;
@@ -4466,6 +4471,8 @@ struct ieee80211_multiple_bssid_configuration {
 
 #define WLAN_OUI_QCN                           0x8CFDF0
 #define WLAN_OUI_QCN_TYPE_1                    1
+#define WLAN_OUI_QCN_HEADER_LEN                4 /* WLAN_OUI_QCN + type */
+#define WLAN_OUI_QCN_ATTR_VER                  1
 #define WLAN_OUI_QCN_ATTR_5GHZ_240MHZ_SUPP     0x0B
 
 #define IEEE80211_EHT_OPER_CTRL_CHANWIDTH_20MHZ        0
diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index 68843b5..3a57a75 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -143,6 +143,9 @@ struct ieee80211_bss {
 
 	/* Keep track of what bits of information we have valid info for. */
 	u8 valid_data;
+
+	struct ieee80211_240mhz_vendor_oper eht_240mhz_capab;
+	u8 eht_240mhz_len;
 };
 
 /**
@@ -2740,8 +2743,11 @@ int ieee80211_put_he_6ghz_cap(struct sk_buff *skb,
 int ieee80211_put_eht_cap(struct sk_buff *skb,
 			  struct ieee80211_sub_if_data *sdata,
 			  const struct ieee80211_supported_band *sband,
-			  const struct ieee80211_conn_settings *conn);
+			  const struct ieee80211_conn_settings *conn,
+			  const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab);
 
+u8 *ieee80211_add_qcn_vendor_ie(struct sk_buff *skb,
+				const struct ieee80211_240mhz_vendor_oper *vendor_ie);
 /* channel management */
 bool ieee80211_chandef_ht_oper(const struct ieee80211_ht_operation *ht_oper,
 			       struct cfg80211_chan_def *chandef);
diff --git a/net/mac80211/mesh.c b/net/mac80211/mesh.c
index 642407f..679b9f6 100644
--- a/net/mac80211/mesh.c
+++ b/net/mac80211/mesh.c
@@ -684,7 +684,7 @@ int mesh_add_eht_cap_ie(struct ieee80211_sub_if_data *sdata,
 	    sdata->vif.bss_conf.chanreq.oper.width == NL80211_CHAN_WIDTH_10)
 		return 0;
 
-	return ieee80211_put_eht_cap(skb, sdata, sband, NULL);
+	return ieee80211_put_eht_cap(skb, sdata, sband, NULL, NULL);
 }
 
 int mesh_add_eht_oper_ie(struct ieee80211_sub_if_data *sdata, struct sk_buff *skb)
diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index c6dfc20..761c9fe 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -166,7 +166,8 @@ ieee80211_determine_ap_chan(struct ieee80211_sub_if_data *sdata,
 			    const struct ieee802_11_elems *elems,
 			    bool ignore_ht_channel_mismatch,
 			    const struct ieee80211_conn_settings *conn,
-			    struct cfg80211_chan_def *chandef)
+			    struct cfg80211_chan_def *chandef,
+			    const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab)
 {
 	const struct ieee80211_ht_operation *ht_oper = elems->ht_operation;
 	const struct ieee80211_vht_operation *vht_oper = elems->vht_operation;
@@ -347,8 +348,7 @@ ieee80211_determine_ap_chan(struct ieee80211_sub_if_data *sdata,
 
 		eht_phy_cap = eht_cap->eht_cap_elem.phy_cap_info[0];
 		he_phy_cap = he_cap->he_cap_elem.phy_cap_info[0];
-		support_320 =
-			eht_phy_cap & IEEE80211_EHT_PHY_CAP0_320MHZ_IN_6GHZ;
+		support_320 = eht_phy_cap & IEEE80211_EHT_PHY_CAP0_320MHZ_IN_6GHZ;
 		support_160 =
 			he_phy_cap &
 			IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_160MHZ_IN_5G;
@@ -375,6 +375,30 @@ ieee80211_determine_ap_chan(struct ieee80211_sub_if_data *sdata,
 		*chandef = eht_chandef;
 	}
 
+	if (eht_oper && eht_240mhz_capab && chandef->width == NL80211_CHAN_WIDTH_160) {
+		struct cfg80211_chan_def eht_chandef = *chandef;
+
+		eht_chandef.width = NL80211_CHAN_WIDTH_320;
+		eht_chandef.center_freq1 =
+			ieee80211_channel_to_frequency(eht_240mhz_capab->ccfs0,
+						       chandef->chan->band);
+
+		eht_chandef.punctured = eht_240mhz_capab->punctured;
+		if (!cfg80211_chandef_valid(&eht_chandef)) {
+			sdata_info(sdata,
+				   "AP EHT 240 MHz information is invalid, disabling EHT\n");
+			return IEEE80211_CONN_MODE_HE;
+		}
+
+		if (!cfg80211_chandef_compatible(chandef, &eht_chandef)) {
+			sdata_info(sdata,
+				   "AP EHT 240 MHz information is incompatible, disabling EHT\n");
+			return IEEE80211_CONN_MODE_HE;
+		}
+
+		*chandef = eht_chandef;
+	}
+
 	return IEEE80211_CONN_MODE_EHT;
 }
 
@@ -945,7 +969,8 @@ again:
 		return ERR_PTR(-ENOMEM);
 
 	ap_mode = ieee80211_determine_ap_chan(sdata, channel, bss->vht_cap_info,
-					      elems, false, conn, ap_chandef);
+					      elems, false, conn, ap_chandef,
+					      elems->eht_240mhz_capab);
 
 	/* this should be impossible since parsing depends on our mode */
 	if (WARN_ON(ap_mode > conn->mode)) {
@@ -1144,7 +1169,7 @@ static int ieee80211_config_bw(struct ieee80211_link_data *link,
 
 	ap_mode = ieee80211_determine_ap_chan(sdata, channel, vht_cap_info,
 					      elems, true, &link->u.mgd.conn,
-					      &ap_chandef);
+					      &ap_chandef, elems->eht_240mhz_capab);
 
 	if (ap_mode != link->u.mgd.conn.mode) {
 		link_info(link,
@@ -1570,6 +1595,7 @@ ieee80211_add_link_elems(struct ieee80211_sub_if_data *sdata,
 {
 	enum nl80211_iftype iftype = ieee80211_vif_type_p2p(&sdata->vif);
 	struct cfg80211_bss *cbss = assoc_data->link[link_id].bss;
+	struct ieee80211_bss *bss = (void *)cbss->priv;
 	struct ieee80211_channel *chan = cbss->channel;
 	const struct ieee80211_sband_iftype_data *iftd;
 	struct ieee80211_local *local = sdata->local;
@@ -1725,7 +1751,8 @@ ieee80211_add_link_elems(struct ieee80211_sub_if_data *sdata,
 
 	if (assoc_data->link[link_id].conn.mode >= IEEE80211_CONN_MODE_EHT)
 		ieee80211_put_eht_cap(skb, sdata, sband,
-				      &assoc_data->link[link_id].conn);
+				      &assoc_data->link[link_id].conn,
+				      bss->eht_240mhz_len ? &bss->eht_240mhz_capab : NULL);
 
 	if (sband->band == NL80211_BAND_S1GHZ) {
 		ieee80211_add_aid_request_ie(sdata, skb);
@@ -2004,11 +2031,14 @@ static int ieee80211_send_assoc(struct ieee80211_sub_if_data *sdata)
 
 	for (link_id = 0; link_id < IEEE80211_MLD_MAX_NUM_LINKS; link_id++) {
 		struct cfg80211_bss *cbss = assoc_data->link[link_id].bss;
+		struct ieee80211_bss *bss = (void *)cbss->priv;
 		size_t elems_len = assoc_data->link[link_id].elems_len;
+		struct ieee80211_supported_band *sband;
 
 		if (!cbss)
 			continue;
 
+		sband = local->hw.wiphy->bands[cbss->channel->band];
 		n_links++;
 
 		size += ieee80211_link_common_elems_size(sdata, iftype, cbss,
@@ -2020,6 +2050,11 @@ static int ieee80211_send_assoc(struct ieee80211_sub_if_data *sdata)
 		/* should be the same across all BSSes */
 		if (cbss->capability & WLAN_CAPABILITY_PRIVACY)
 			capab |= WLAN_CAPABILITY_PRIVACY;
+
+		if (bss->eht_240mhz_len && sband->band == NL80211_BAND_5GHZ)
+			size += (2 + WLAN_OUI_QCN_HEADER_LEN +
+				 2 + sizeof(struct ieee80211_qcn_version_vendor_elem) +
+				 2 + sizeof(struct ieee80211_240mhz_vendor_oper));
 	}
 
 	if (ieee80211_vif_is_mld(&sdata->vif)) {
diff --git a/net/mac80211/parse.c b/net/mac80211/parse.c
index d5d4678..9ae443f 100644
--- a/net/mac80211/parse.c
+++ b/net/mac80211/parse.c
@@ -288,7 +288,7 @@ static void ieee802_11_parse_elems_vendor_qcn(const u8 *pos, u8 elen,
 		switch (pos[0]) {
 		case WLAN_OUI_QCN_ATTR_5GHZ_240MHZ_SUPP:
 			if (rem_elen >= 11) {
-				elems->eht_240mhz_capab = pos + 2;
+				elems->eht_240mhz_capab = (const struct ieee80211_240mhz_vendor_oper *)pos + 2;
 				elems->eht_240mhz_len = pos[1];
 			}
 			break;
diff --git a/net/mac80211/scan.c b/net/mac80211/scan.c
index a64fd3f..d6108c7 100644
--- a/net/mac80211/scan.c
+++ b/net/mac80211/scan.c
@@ -157,6 +157,13 @@ void ieee80211_inform_bss(struct wiphy *wiphy,
 	else
 		bss->vht_cap_info = 0;
 
+	if (cbss->channel->band == NL80211_BAND_5GHZ &&
+	    elems->eht_240mhz_len && elems->eht_240mhz_capab) {
+		bss->eht_240mhz_len = elems->eht_240mhz_len;
+		memcpy(&bss->eht_240mhz_capab, elems->eht_240mhz_capab,
+		       sizeof(struct ieee80211_240mhz_vendor_oper));
+	}
+
 	kfree(elems);
 }
 
diff --git a/net/mac80211/tdls.c b/net/mac80211/tdls.c
index d94bf93..c0179a3 100644
--- a/net/mac80211/tdls.c
+++ b/net/mac80211/tdls.c
@@ -578,7 +578,7 @@ ieee80211_tdls_add_setup_start_ies(struct ieee80211_link_data *link,
 	if (action_code == WLAN_TDLS_SETUP_REQUEST ||
 	    action_code == WLAN_TDLS_SETUP_RESPONSE ||
 	    action_code == WLAN_PUB_ACTION_TDLS_DISCOVER_RES)
-		ieee80211_put_eht_cap(skb, sdata, sband, NULL);
+		ieee80211_put_eht_cap(skb, sdata, sband, NULL, NULL);
 
 	/* add any remaining IEs */
 	if (extra_ies_len) {
diff --git a/net/mac80211/util.c b/net/mac80211/util.c
index a5f758e..1752b5a 100644
--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -1571,7 +1571,7 @@ static int ieee80211_put_preq_ies_band(struct sk_buff *skb,
 	if (cfg80211_any_usable_channels(local->hw.wiphy, BIT(sband->band),
 					 IEEE80211_CHAN_NO_HE |
 					 IEEE80211_CHAN_NO_EHT)) {
-		err = ieee80211_put_eht_cap(skb, sdata, sband, NULL);
+		err = ieee80211_put_eht_cap(skb, sdata, sband, NULL, NULL);
 		if (err)
 			return err;
 	}
@@ -5009,6 +5009,50 @@ u8 *ieee80211_add_wmm_info_ie(u8 *buf, u8 qosinfo)
 	return buf;
 }
 
+static u8 *ieee80211_add_qcn_version_vendor_ie(u8 *buf)
+{
+	*buf++ = WLAN_OUI_QCN_ATTR_VER;
+	*buf++ = 2; /* length */
+	*buf++ = 1; /* version */
+	*buf++ = 0; /* sub-version */
+	return buf;
+}
+
+static u8 *ieee80211_add_240mhz_vendor_ie(u8 *buf,
+					  const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab)
+{
+	*buf++ = WLAN_OUI_QCN_ATTR_5GHZ_240MHZ_SUPP;
+	*buf++ = sizeof(struct ieee80211_240mhz_vendor_oper);
+	memcpy(buf, (u8 *) eht_240mhz_capab, sizeof(struct ieee80211_240mhz_vendor_oper));
+	buf += sizeof(struct ieee80211_240mhz_vendor_oper);
+	return buf;
+}
+
+u8 *ieee80211_add_qcn_vendor_ie(struct sk_buff *skb,
+				const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab)
+{
+	size_t len = WLAN_OUI_QCN_HEADER_LEN;
+	u8 *pos;
+
+	len += (2 + sizeof(struct ieee80211_qcn_version_vendor_elem));
+
+	if (eht_240mhz_capab)
+		len += (2 + sizeof(struct ieee80211_240mhz_vendor_oper));
+
+	pos = skb_put(skb, 2 + len);
+	*pos++ = WLAN_EID_VENDOR_SPECIFIC;
+	*pos++ = len;
+	*pos++ = 0x8c; /* WLAN_OUI_QCN */
+	*pos++ = 0xfd;
+	*pos++ = 0xf0;
+	*pos++ = WLAN_OUI_QCN_TYPE_1;
+
+	pos = ieee80211_add_qcn_version_vendor_ie(pos);
+	pos = ieee80211_add_240mhz_vendor_ie(pos, eht_240mhz_capab);
+
+	return pos;
+}
+
 void ieee80211_txq_get_depth(struct ieee80211_txq *txq,
 			     unsigned long *frame_cnt,
 			     unsigned long *byte_cnt)
@@ -5091,7 +5135,8 @@ u8 ieee80211_ie_len_eht_cap(struct ieee80211_sub_if_data *sdata)
 int ieee80211_put_eht_cap(struct sk_buff *skb,
 			  struct ieee80211_sub_if_data *sdata,
 			  const struct ieee80211_supported_band *sband,
-			  const struct ieee80211_conn_settings *conn)
+			  const struct ieee80211_conn_settings *conn,
+			  const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab)
 {
 	const struct ieee80211_sta_he_cap *he_cap =
 		ieee80211_get_he_iftype_cap_vif(sband, &sdata->vif);
@@ -5100,7 +5145,7 @@ int ieee80211_put_eht_cap(struct sk_buff *skb,
 	bool for_ap = sdata->vif.type == NL80211_IFTYPE_AP;
 	struct ieee80211_eht_cap_elem_fixed fixed;
 	struct ieee80211_he_cap_elem he;
-	u8 mcs_nss_len, ppet_len;
+	u8 mcs_nss_len, ppet_len, *pos;;
 	u8 orig_mcs_nss_len;
 	u8 ie_len;
 
@@ -5178,6 +5223,9 @@ int ieee80211_put_eht_cap(struct sk_buff *skb,
 	if (ppet_len)
 		skb_put_data(skb, &eht_cap->eht_ppe_thres, ppet_len);
 
+	if (eht_240mhz_capab && sband->band == NL80211_BAND_5GHZ)
+		pos = ieee80211_add_qcn_vendor_ie(skb, eht_240mhz_capab);
+
 	return 0;
 }
 
-- 
2.34.1

